src/Perl6/Metamodel/Concretization.nqp

```diff
@@ -1,36 +1,84 @@
 # Support for mapping of non-specialized roles into their concretized state.
 role Perl6::Metamodel::Concretization {
     has @!concretizations;
+    has %!conc_table;
 
     method add_concretization($obj, $role, $concrete) {
         @!concretizations[+@!concretizations] := [$role, $concrete];
+        nqp::scwbdisable();
+        %!conc_table := nqp::hash(); # reset the cache
+        nqp::scwbenable();
     }
 
-    method concretization_lookup($obj, $ptype, :$local, :$transitive) {
+    method concretizations($obj, :$local = 0, :$transitive = 1) {
+        my @conc;
+        for @!concretizations {
+            my @c := $transitive ?? [] !! @conc;
+            nqp::push(@c, $_[1]);
+            if $transitive && nqp::can($_[1].HOW, 'concretizations') {
+                for $_[1].HOW.concretizations($_[1], :$local) {
+                    nqp::push(@c, $_);
+                }
+            }
+            nqp::push(@conc, @c) if $transitive;
+        }
+        @conc := self.c3_merge(@conc) if $transitive;
+        unless $local {
+            for self.parents($obj, :local) {
+                for $_.HOW.concretizations($_, :$local, :$transitive) {
+                    nqp::push(@conc, $_)
+                }
+            }
+        }
+        @conc
+    }
+
+    method !rebuild_table() {
+        for @!concretizations {
+            nqp::scwbdisable();
+            %!conc_table{~nqp::objectid(nqp::decont($_[0]))} := nqp::decont($_[1]);
+            nqp::scwbenable();
+        }
+    }
+
+    method concretization_lookup($obj, $ptype, :$local = 0, :$transitive = 1, :$relaxed = 0) {
+        self.'!rebuild_table'() if nqp::elems(%!conc_table) < nqp::elems(@!concretizations);
+        return [0] unless !$local || $transitive || nqp::elems(%!conc_table);
+        $ptype := nqp::decont($ptype);
+        return [0] if $obj.HOW.is_composed($obj) && !nqp::istype($obj, $ptype);
+        my $id := ~nqp::objectid($ptype);
         my @result;
-        if nqp::istype($obj, $ptype) {
+        if nqp::existskey(%!conc_table, $id) {
+            return [1, %!conc_table{$id}];
+        }
+        if $relaxed {
+            # Try search by role group for curryings. The first match is ok. Used by FQN method calls.
             for @!concretizations {
-                if nqp::decont($_[0]) =:= nqp::decont($ptype) {
+                next unless $_[0].HOW ~~ Perl6::Metamodel::CurriedRoleHOW;
+                if nqp::decont($_[0].HOW.curried_role($_[0])) =:= $ptype {
```

Comment:
Can this check be written in a way that would allow people to use this meta-role with their own HOWs for parameterized types like it did before?

Reply:
Any suggestions? Code which would be broken with this change?

Reply:
> Code which would be broken with this change?

This would make it so I can't use this meta-role with one of the HOWs in a typeclass library I've been working on for a while now (which used to be a monad library). While it's not something I think will be in a releasable state for quite a while yet, if at all, it's useful for testing how you'd go about implementing some behaviours related to parameterization whose syntax is supported, but none of the parametric types in Rakudo know how to deal with. Should they be behaviours that should be supported, there would be less risk involved in implementing them for Rakudo's parametric types this way since the solutions will have already been tested through this library.

> Any suggestions?

What I'd do is narrow down the types that were concretized to types that are both parametric and inheritalizable, then check if the type parameterized is the parametric type passed:

```perl6
next unless $_[0].HOW.archetypes.parametric && $_[0].HOW.archetypes.inheritalizable;
if nqp::typeparameterized($_[0]) =:= $ptype { ... }
```

While types that aren't parameterized but fit these archetypes (like `Perl6::Metamodel::ParametricRoleHOW`) would pass the first check if they ended up here, `nqp::typeparameterized` is not one of the parametric type ops that throws if it doesn't receive a type that can actually be used with it.

Reply:
Actually, now that I think about it, there may be ways I can handle typeclass instances without needing to make the typeclass a parent of the types they mark as instances, so I may not need this meta-role in the end. Still, someone may need to make a punnable parametric HOW some day.

Reply:
@Kaiepi I wasn't able to commit my yesterday's change, but now it looks like this:

```
                next unless $_[0].HOW.archetypes.parametric;
                my $conc := nqp::can($_[0].HOW, 'curried_role')
                                ?? $_[0].HOW.curried_role($_[0])
                                !! $_[0];
                if nqp::decont($conc) =:= $ptype {
```

So, eventually, it should be fine with custom `RoleHOW`s and one could even simulate curryings if desire.

Reply:
@Kaiepi I think my code would fulfill your requirements.